pkg/fetcher/definition.go

```diff
@@ -22,33 +23,71 @@ type Definition struct {
 	reader client.Reader
 }
 
+type uniqueTestDefinitions map[types.UID]v1alpha1.TestDefinition
+
 func (s *Definition) FindMatching(suite v1alpha1.ClusterTestSuite) ([]v1alpha1.TestDefinition, error) {
 	ctx := context.TODO()
-	if len(suite.Spec.Selectors.MatchNames) > 0 {
-		return s.findByNames(ctx, suite)
+	acc := make(uniqueTestDefinitions)
+
+	err := s.findByNames(ctx, suite, acc)
+	if err != nil {
+		return nil, err
 	}
-	// TODO(aszecowka) later so far we return all test definitions for all namespaces (https://github.com/kyma-incubator/octopus/issues/7)
+
+	err = s.findByLabelExpressions(ctx, suite, acc)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(acc) > 0 {
+		return acc.getValues(), nil
+	}
+
 	var list v1alpha1.TestDefinitionList
 	if err := s.reader.List(ctx, &client.ListOptions{Namespace: ""}, &list); err != nil {
 		return nil, errors.Wrap(err, "while listing test definitions")
 	}
 	return list.Items, nil
 }
 
-func (s *Definition) findByNames(ctx context.Context, suite v1alpha1.ClusterTestSuite) ([]v1alpha1.TestDefinition, error) {
-	var list []v1alpha1.TestDefinition
+func (s *Definition) findByNames(ctx context.Context, suite v1alpha1.ClusterTestSuite, acc uniqueTestDefinitions) error {
 	for _, tRef := range suite.Spec.Selectors.MatchNames {
 		def := v1alpha1.TestDefinition{}
 		err := s.reader.Get(ctx, types.NamespacedName{Name: tRef.Name, Namespace: tRef.Namespace}, &def)
 		wrappedErr := errors.Wrapf(err, "while fetching test definition from selector [name: %s, namespace: %s]", tRef.Name, tRef.Namespace)
 		switch {
 		case err == nil:
 		case k8serrors.IsNotFound(err):
-			return nil, humanerr.NewError(wrappedErr, fmt.Sprintf("Test Definition [name: %s, namespace: %s] does not exist", tRef.Name, tRef.Namespace))
+			return humanerr.NewError(wrappedErr, fmt.Sprintf("Test Definition [name: %s, namespace: %s] does not exist", tRef.Name, tRef.Namespace))
 		default:
-			return nil, humanerr.NewError(wrappedErr, "Internal error")
+			return humanerr.NewError(wrappedErr, "Internal error")
+		}
+		acc[def.UID] = def
+	}
+	return nil
+}
+
+func (s *Definition) findByLabelExpressions(ctx context.Context, suite v1alpha1.ClusterTestSuite, acc uniqueTestDefinitions) error {
+	for _, expr := range suite.Spec.Selectors.MatchLabelExpressions {
+		selector, err := labels.Parse(expr)
+		if err != nil {
+			return errors.Wrapf(err, "while parsing label expression [expression: %s]", expr)
+		}
+		var list v1alpha1.TestDefinitionList
+		if err := s.reader.List(ctx, &client.ListOptions{LabelSelector: selector}, &list); err != nil {
```

Comment:
this line is not tested, I've changed it to: "if err := s.reader.List(ctx, &client.ListOptions{}, &list); err != nil {" and still the test is green. This is important part of the matching logic.
Consider filtering directly by fake client, check the controller runtime version (see https://github.com/kubernetes-sigs/controller-runtime/issues/293)

Reply:
It's not available in our controller-runtime version and we cannot use newer as it requires apimachinery 1.13

Reply:
I changed the mock to use the LAbelSelector. Now the logic is checked by the test.